### PR TITLE
🎨 Payments-methods logic in `payments` service

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

Implements payment-methods logic in the `payments` service. The logic has been basically copied and adapted from that existing in the `webserver.payments` plugin. Nonetheless, the code in the `webserver` will still fake the payment-method until app-motion's app-gateway fully implements the payment-method infrastructure (https://github.com/ITISFoundation/appmotion-exchange/issues/11)

### Highlights
payments

- ✨`db/payments_methods_repo`: implemented a repository layer for payment-methods
- ✨`api/rest/_acknowledgements` : implemented the rest interface to ACK payment-methods
- ✨`api/rpc/_payments_methods` : implements rpc interface for payment-methods exposed to the web-server
- ✨`services/payments_gateway` : implements the http client calls to the  payment-gateway API for payment-methods
- ♻️ minor models rename, tests etc

## Related issue/s

- Moves logic to payment service in #4751

## How to test

Driving tests
```
cd services/payments
make install-dev
pytest -vv tests/unit
```

## DevOps 

- None

### CODE REVIEW

File: packages/models-library/src/models_library/api_schemas_webserver/wallets.py [hunk 1]
```diff
@@ -59,10 +59,12 @@ class CreateWalletPayment(InputSchema):
     comment: str = FieldNotRequired(max_length=100)
 
 
-class WalletPaymentCreated(OutputSchema):
+class WalletPaymentInitiated(OutputSchema):
     payment_id: PaymentID
-    payment_form_url: HttpUrl = Field(
-        ..., description="Link to external site that holds the payment submission form"
+    payment_form_url: HttpUrl | None = Field(
+        None,
```
Comment (sanderegg) [inline packages/models-library/src/models_library/api_schemas_webserver/wallets.py:65]:
```suggestion
        default=None,
```

File: services/payments/src/simcore_service_payments/api/rest/_acknowledgements.py [hunk 2]
```diff
@@ -3,15 +3,17 @@
 
 from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, status
 from servicelib.logging_utils import log_context
+from simcore_postgres_database.models.payments_methods import InitPromptAckFlowState
 from simcore_postgres_database.models.payments_transactions import (
     PaymentTransactionState,
 )
+from simcore_service_payments.db.payments_methods_repo import PaymentsMethodsRepo
```
Comment (sanderegg) [inline services/payments/src/simcore_service_payments/api/rest/_acknowledgements.py:10]: relative import?

File: services/payments/src/simcore_service_payments/db/base.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        assert db_engine is not None  # nosec
+        self.db_engine = db_engine
+
+    # FIXME:
+    # async def transaction(self, connection: None) -> Iterable[]:
+    #     if connection is not None:
+    #         yield connection
+    #     else:
+    #         with self.db_engine.begin() as conn:
+    #             yield conn
```
Comment (sanderegg) [inline services/payments/src/simcore_service_payments/db/base.py:19]: forgotten?

File: services/payments/tests/unit/test_services_payments_gateway.py [hunk 4]
```diff
@@ -116,3 +123,84 @@ async def test_one_time_payment_workflow(
     if mock_payments_gateway_service_or_none:
         assert mock_payments_gateway_service_or_none.routes["init_payment"].called
         assert mock_payments_gateway_service_or_none.routes["cancel_payment"].called
+
+
+@pytest.mark.testit()
```
Comment (sanderegg) [inline services/payments/tests/unit/test_services_payments_gateway.py:128]: forgotten?

**Review by sanderegg — APPROVED** [review]
this is looking good. I am not sure I understand how you distringuish between  log.debug messages, todos and fixme though.

File: services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+    # Prevents card from being used
+    repo = PaymentsMethodsRepo(db_engine=app.state.engine)
+
+    await repo.update_ack_payment_method(
+        payment_method_id,
+        completion_state=InitPromptAckFlowState.CANCELED,
+        state_message="User cancelled",
+    )
+    # TODO: Notify?
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py:113]: was this forgotten?

File: services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+@router.expose()
+async def cancel_creation_of_payment_method(
+    app: FastAPI,
+    *,
+    payment_method_id: PaymentMethodID,
+    user_id: UserID,
+    wallet_id: WalletID,
+) -> None:
+    # TODO: with db transaction
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py:103]: was this forgotten as well?

File: services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    user_id: UserID,
+    wallet_id: WalletID,
+) -> PaymentMethodGet:
+
+    repo = PaymentsMethodsRepo(db_engine=app.state.engine)
+    acked = await repo.get_payment_method(
+        payment_method_id, user_id=user_id, wallet_id=wallet_id
+    )
+
+    #
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py:185]: I'd remove it or update it, not that useful being an empty comment

File: services/payments/src/simcore_service_payments/db/payments_methods_repo.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    async def insert_payment_method(
+        self,
+        payment_method_id: PaymentMethodID,
+        *,
+        user_id: UserID,
+        wallet_id: WalletID,
+        completion_state: InitPromptAckFlowState,
+        state_message: str | None,
+    ) -> PaymentsMethodsDB:
+        # TODO: how to do this on a single begin()?
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/db/payments_methods_repo.py:105]: was this one forgotten as well?

**Review by GitHK — APPROVED** [review]
👍

File: services/payments/src/simcore_service_payments/api/rest/_acknowledgements.py (hunk 2 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/api/rest/_acknowledgements.py:10]: thx. that is the auto importer. I always forget to check.

File: services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py (hunk 7 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py:185]: what do you mean? This is not a comment but a `datetime | None`
Or perhaps you mean the `#` .. ok yes

File: services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py (hunk 6 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py:103]: Very very sorry ... i should not PR at night!

File: services/payments/src/simcore_service_payments/db/payments_methods_repo.py (hunk 8 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/db/payments_methods_repo.py:105]: sorry again

File: services/payments/src/simcore_service_payments/db/base.py (hunk 3 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/db/base.py:19]: this is embarrasing... 
last time I PR in the evening
again sorry!

File: packages/models-library/src/models_library/api_schemas_webserver/wallets.py [hunk 9]
```diff
@@ -59,10 +59,12 @@ class CreateWalletPayment(InputSchema):
...
 
-class WalletPaymentCreated(OutputSchema):
+class WalletPaymentInitiated(OutputSchema):
     payment_id: PaymentID
-    payment_form_url: HttpUrl = Field(
-        ..., description="Link to external site that holds the payment submission form"
+    payment_form_url: HttpUrl | None = Field(
+        default=None,
+        description="Link to external site that holds the payment submission form."
+        "None if no prompt step is required (e.g. pre-selected credit card)",
```
Comment (bisgaard-itis) [inline packages/models-library/src/models_library/api_schemas_webserver/wallets.py:67]:
```suggestion
        description="Link to external site that holds the payment submission form. "
        "None if no prompt step is required (e.g. pre-selected credit card)",
```

**Review by bisgaard-itis — APPROVED** [review]
Nice! Looks good to me

**Review by matusdrobuliak66 — APPROVED** [review]
:1st_place_medal:

**Pull request merged**